agent/consul/acl_server.go

```diff
@@ -150,7 +150,12 @@ func (s *Server) ResolveIdentityFromToken(token string) (bool, structs.ACLIdenti
 		return true, aclToken, nil
 	}
 
-	return s.InPrimaryDatacenter() || index > 0, nil, acl.ErrNotFound
+	defaultErr := acl.ErrNotFound
+	canBootstrap, _, _ := s.fsm.State().CanBootstrapACLToken()
+	if canBootstrap {
+		defaultErr = fmt.Errorf("ACL system must be bootstrapped before making any requests that require authorization: %w", defaultErr)
```

Comment:
non-blocking: When I tested this locally, I only saw this error when I provide a token (i.e. if I do not use the anonymous token):

```
$ CONSUL_HTTP_TOKEN=root ./bin/consul acl policy list
Failed to retrieve the policy list: Unexpected response code: 403 (ACL system must be bootstrapped before making any requests that require authorization: ACL not found)
$ ./bin/consul acl policy list
Failed to retrieve the policy list: Unexpected response code: 403 (Permission denied: anonymous token lacks permission 'acl:read'. The anonymous token is used implicitly when a request does not specify a token.)
```

If they haven't yet ACL bootstrapped, then in order to see the `ACL system must be bootstrapped` message, would they have to pass some token that has not yet been created (maybe a random value)? If so, it seems unlikely someone would pass a random token value. So I feel like this would be more valuable if the `ACL system must be bootstrapped` error messages is returned in the case of the anonymous token as well.

Reply:
Did you test this on a single server with the -bootstrap flag?

The anonymous token gets created when a server acquires leadership (https://github.com/hashicorp/consul/blob/e91bc9c0586c5cd11b933f854165189720969340/agent/consul/leader.go#L501), which is intended so to avoid downtime / loss of functionality during the ACL enablement process. It's a bit of a tradeoff between providing a useful error and allowing some operations to continue. In fact, there is a related task for allowing anonymous token synthesis for the case where default_policy=allow to expand the ability to use the anonymous token for basic operations before leadership has been established: https://github.com/hashicorp/consul/pull/16200

In multi-server clusters, you will get this bootstrap error before the servers can be joined when you have a default deny policy. The bootstrap error generally appears a lot more with default_policy=deny, because the anonymous token becomes almost useless in that case.

It isn't great UX, but when users get the permission denied error for the anonymous token, if they try to use any non-nil / non-anonymous value as a next step, they will then get this bootstrap error which should point them in the right direction if they want to be able to create more tokens.

Reply:
Ah.

> Did you test this on a single server with the -bootstrap flag?

Yes, though the behavior is the same with a multi-server cluster with -bootstrap-expect.

> In multi-server clusters, you will get this bootstrap error before the servers can be joined when you have a default deny policy.

Except if the requests get stuck in RPC forwarding, they hang trying to forward to the leader. But the Agent API requests that don't do RPC forwarding do show the error.

* `consul acl policy list` (for example) will hang during RPC forwarding to the leader when there's no cluster leader.
 
  ```
  $ CONSUL_HTTP_TOKEN=root consul acl policy list
  Failed to retrieve the policy list: Unexpected response code: 500 (No cluster leader)
  ```

* On the other hand, `consul members` and `consul join` will show the `ACL system must be bootstrapped` error when there is no cluster leader.

  ```
  $ consul members
  Error retrieving members: Unexpected response code: 403 (ACL system must be bootstrapped before making any 
  requests that require authorization: ACL not found)
  $ consul join 127.0.0.1:9300
  Error joining address '127.0.0.1:9300': Unexpected response code: 403 (ACL system must be bootstrapped before 
  making any requests that require authorization: ACL not found)
  Failed to join any nodes.
  ```

Reply:
In any case, this is better than before 🙂 

I do think it's not likely people will try to use random token values that they haven't created yet, but seeing these errors on the `consul members` and `consul join` commands is super useful.

Reply:
> Yes, though the behavior is the same with a multi-server cluster with -bootstrap-expect.

Oh, no. I see now. Ignore this! The multi-server case never elects a leader, so no anonymous token versus the single server case (pebkac problem where I was still using -bootstrap).